src/launcher/java/org/truffleruby/launcher/RubyLauncher.java

```diff
@@ -13,27 +13,104 @@
 import org.graalvm.launcher.AbstractLanguageLauncher;
 import org.graalvm.options.OptionCategory;
 import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.PolyglotException;
+import org.graalvm.polyglot.Source;
+import org.truffleruby.launcher.options.CommandLineException;
 import org.truffleruby.launcher.options.CommandLineOptions;
+import org.truffleruby.launcher.options.CommandLineParser;
+import org.truffleruby.launcher.options.ExecutionAction;
 import org.truffleruby.launcher.options.OptionsCatalog;
 
+import java.io.File;
+import java.io.PrintStream;
+import java.lang.management.ManagementFactory;
+import java.lang.reflect.Method;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
 public class RubyLauncher extends AbstractLanguageLauncher {
 
+    public static final String LANGUAGE_ID = "ruby";
+    public static final String ENGINE_ID = "truffleruby";
+    public static final String LANGUAGE_VERSION = "2.3.5";
+    public static final int LANGUAGE_REVISION = 59905;
+    public static final String BOOT_SOURCE_NAME = "main_boot_source";
+    public static final String RUBY_COPYRIGHT = "truffleruby - Copyright (c) 2013-2017 Oracle and/or its affiliates";
+
+    // Properties set directly on the java command-line with -D for image building
+    public static final String LIBSULONG_DIR = isAOT() ? System.getProperty("truffleruby.native.libsulong_dir") : null;
+    public static final boolean PRE_INITIALIZE_CONTEXTS = System.getProperty("polyglot.engine.PreinitializeContexts") != null;
+
     private final CommandLineOptions config = new CommandLineOptions();
 
     public static void main(String[] args) {
         new RubyLauncher().launch(args);
     }
 
+    // TODO (pitr-ch 22-Feb-2018): replace with a call to sdk API when available
+    static boolean isGraal() {
+        final CommandLineOptions config = new CommandLineOptions();
+        config.setOption(OptionsCatalog.GRAAL_WARNING_UNLESS, false);
+        config.setOption(OptionsCatalog.POST_BOOT, false);
+        config.setOption(OptionsCatalog.NO_HOME_PROVIDED, true);
+
+        try (Context context = RubyLauncher.createContext(Context.newBuilder(), config)) {
+            final Source source = Source.newBuilder(
+                    LANGUAGE_ID,
+                    // language=ruby
+                    "Truffle.graal?",
+                    BOOT_SOURCE_NAME).internal(true).buildLiteral();
+            return context.eval(source).asBoolean();
+        } catch (PolyglotException e) {
+            System.err.println("truffleruby: " + e.getMessage());
+            e.printStackTrace();
+            return false;
+        }
+    }
```

Comment:
This seems a very expensive cost on startup.
Does it mean we will create 2 RubyContext, initialize the core library twice, etc if using `ruby -ve 'p 3'`?
I think we need to address this before merging this PR.

Maybe we can simply handle this in the Bash launcher for now (e.g., checking if given a Graal jar), or push for this to be added in the SDK as soon as possible.

Reply:
Yeah I'm not sure about this part either.

Reply:
It's a workaround until there is a API for it in sdk (issue created), until then I think the overhead for only `-v` option is fine.

Reply:
I don't think we would want to ship a GraalVM where `ruby -v` takes a full startup or
`ruby -ve "p :hi"` takes 2 times the startup on JVM.
So I think unless there is a requirement to merge this PR before that justifies this risk, we need to find a workaround or push and wait for the fix in Truffle (or do it ourselves).

Reply:
Related, I found this today:

```
$ graalvm-0.31/Contents/Home/bin/ruby -v -e 'loop do { }'
truffleruby 0.31, like ruby 2.3.5 <GraalVM 0.31 1.8.0_151-b12 with Graal> [darwin-x86_64]
truffleruby: -e:1: syntax error, unexpected end-of-file
loop do { } (SyntaxError)
-e:1: syntax error, unexpected end-of-file
loop do { } (SyntaxError)
```

Reply:
Is there anything where slow `-v` would be causing problems?

Reply:
Extremely bad user experience is what I refer to above.

We cannot easily know if users use `ruby -v` or `ruby -v ...` often but I expect it's far from non-existent (example: I always use `ruby -v` to show the version).
`-v` also seems used for benchmarks: https://bugs.ruby-lang.org/issues/14462
MSpec used to run with `-v` to show the Ruby being tested, it's a common/convenient way to report which implementation is being run.

Reply:
Hm does sound that harmless as I've thought. I'll add a shortcut for GraalVM we can assume that Graal is always present there. I think the temporary slowdown should be fine outside of GraalVM.

Reply:
@chrisseaton isn't there an extra `do`?

Reply:
Yes it’s a syntax error. But the point is it’s printed twice. That’s not right.

Reply:
Ah sorry missed that.

Reply:
@chrisseaton That issue is totally unrelated though, let's create a separate issue.

Reply:
> I'll add a shortcut for GraalVM we can assume that Graal is always present there. I think the temporary slowdown should be fine outside of GraalVM.

That's a first step but there are other distributions and other ways to run TruffleRuby outside GraalVM and I think in none of the cases it's acceptable to create an extra RubyContext which has many side effects and wastes a lot of resources (including potentially loading native libraries multiple times, creating a lot of work for GC, delaying startup by seconds on JVM, etc).

> It's a workaround until there is a API for it in sdk (issue created)

What's the issue number?

How can the sdk know if Graal is available for Truffle compilations?
It seems rather complicated, I see 3 scenarios:
* GraalVM: Graal always available.
* SVM (native): Graal always available (not technically a requirement but we never build SVM images without Graal so good enough).
* The rest: we should be able to classload TruffleRuntime dynamically to figure it out, since everything is on the bootclasspath.

I suspect this is something we can implement in this PR rather easily, and is not a huge hack creating an extra RubyContext just to get a boolean out for something that is not Ruby-specific.

Reply:
The issue is GR-8367. I am not sure how will the issue be fixed. I had a look at the dynamic Truffle loading and it works well without too much effort, pushing.